File: exercism/haskell/list-ops/src/ListOps.hs
module ListOps
  ( length
  , reverse
  , map
  , filter
  , foldr
  , foldl'
  , (++)
  , concat
  ) where

import Prelude hiding
  ( length, reverse, map, filter, foldr, (++), concat )

-- Hutton, "A tutorial on the universality and expressiveness of fold"
-- http://www.cs.nott.ac.uk/~pszgmh/fold.pdf

foldl' :: (b -> a -> b) -> b -> [a] -> b
foldl' _ z []     = z
foldl' f z (x:xs) = let z' = f z x in z' `seq` foldl' f z' xs

foldr :: (a -> b -> b) -> b -> [a] -> b
foldr _ z []     = z
foldr f z (x:xs) = f x $ foldr f z xs

length :: [a] -> Int
length = foldl' (+) 0 . map (const 1)

reverse :: [a] -> [a]
reverse = foldl' (flip (:)) []

map :: (a -> b) -> [a] -> [b]
map f = foldr (\x xs -> f x : xs) []

filter :: (a -> Bool) -> [a] -> [a]
filter p = foldr (\x xs -> if p x then x : xs else xs) []

(++) :: [a] -> [a] -> [a]
xs ++ ys = foldr (:) ys xs

concat :: [[a]] -> [a]
concat = foldl' (++) []

File: exercism/ocaml/list-ops/list_ops.ml
let rec fold ~init ~f = function
  | [] -> init
  | x :: xs -> fold ~init:(f init x) ~f xs

let length xs = fold ~init:0 ~f:(fun n _ -> succ n) xs
let reverse xs = fold ~init:[] ~f:(Fun.flip List.cons) xs
let map ~f xs = fold ~init:[] ~f:(fun ys x -> f x :: ys) (reverse xs)

let filter ~f xs =
  fold ~init:[] ~f:(fun ys x -> if f x then x :: ys else ys) (reverse xs)

let append xs ys = fold ~init:ys ~f:(Fun.flip List.cons) (reverse xs)
let concat xss = fold ~init:[] ~f:append xss
